lib/Storage/Utils.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/SmallString.h"
+
+#include "revng/Support/PathList.h"
+
+// Utility function that removes any attempt to introduce parent notation in a
+// path. This is used in the StorageClient implementations to avoid escaping
+// outside the "root" directory. This function will produce the same output
+// ('foo/bar') given any of these:
+// * foo/bar
+// * ./foo/../foo/bar
+// * /foo/bar
+// * ../../foo/bar
```

Comment:
These two are weird, I'd either: resolve the path and ensure it's below the root or simply abort if the path starts with `/`, ends with `/..` or contains `../`.

I'd prefer the latter, unless we use this for command line arguments too. Then the user might actually go with `..`, then we can go with the former option.

Reply:
> unless we use this for command line arguments too

we do

> resolve the path and ensure it's below the root

That's by design of the `normalizePath` function: the output path is guaranteed to be below the root

Reply:
Yes, but handling `../foo` as `foo` is weird.
Let's make the path absolute, check if it's below the root and if it's not fail instead of succeeding in unexpected ways.

Reply:
Changed how things work:
* `fromLocalStorage` will compute the absolute path before creating the `{File|Directory}Path`, so on that side it's guaranteed that we don't get `../`
* Changed the logic to do `revng_assert(checkPath(Path))` where checkPath returns false when it detects `../`, `./`, `//` or leading `/`